openvino_xai/methods/white_box/torch.py

```diff
@@ -0,0 +1,257 @@
+        return {
+            "prediction": output,
+            SALIENCY_MAP_OUTPUT_NAME: torch.empty_like(output),
+        }
+
+    def generate_saliency_map(self, data: np.ndarray) -> np.ndarray:
+        """Return saliency map."""
+        model_output = self.model_forward(data)
+        return model_output[SALIENCY_MAP_OUTPUT_NAME]
+
+    @staticmethod
+    def _normalize_map(saliency_map: torch.Tensor) -> torch.Tensor:
+        """Normalize saliency maps."""
+        max_values = saliency_map.max(dim=-1, keepdim=True).values
+        min_values = saliency_map.min(dim=-1, keepdim=True).values
+        saliency_map = 255 * (saliency_map - min_values) / (max_values - min_values + 1e-12)
+        return saliency_map.to(torch.uint8)
+
+
+class ActivationMap(TorchMethod):
+    """ActivationMap. Mean of the feature map along the channel dimension."""
+
+    def _output_hook(self, module: torch.nn.Module, inputs: Any, output: torch.Tensor) -> Dict[str, torch.Tensor]:
+        feature_map = self._feature_map
+        batch_size, _, h, w = feature_map.shape
+        activation_map = torch.mean(feature_map, dim=1)
+        if self._embed_scaling:
+            activation_map = activation_map.reshape((batch_size, h * w))
+            activation_map = self._normalize_map(activation_map)
+            activation_map = activation_map.reshape((batch_size, h, w))
+        return {
+            "prediction": output,
+            SALIENCY_MAP_OUTPUT_NAME: activation_map,
+        }
+
+
+class ReciproCAM(TorchMethod):
+    """Implementation of Recipro-CAM for class-wise saliency map.
+
+    Recipro-CAM: gradient-free reciprocal class activation map (https://arxiv.org/pdf/2209.14074.pdf)
+
+    :param optimize_gap: Whether to optimize out Global Average Pooling operation
+    :type optimizae_gap: bool
+    """
+
+    def __init__(self, *args, optimize_gap: bool = False, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._optimize_gap = optimize_gap
+
+    def _feature_hook(self, module: torch.nn.Module, inputs: Any, output: torch.Tensor) -> torch.Tensor:
+        """feature_maps -> vertical stack of feature_maps + mosaic_feature_maps."""
+        batch_size, c, h, w = self._feature_shape = output.shape
+        feature_map = output
+        if self._optimize_gap:
+            feature_map = feature_map.reshape([batch_size, c, h * w]).mean(dim=-1)[:, :, None, None]  # Spatial average
+        feature_maps = [feature_map]
+        for i in range(batch_size):
+            mosaic_feature_map = self._get_mosaic_feature_map(output[i], c, h, w)
+            feature_maps.append(mosaic_feature_map)
+        return torch.cat(feature_maps)
+
+    def _output_hook(self, module: torch.nn.Module, inputs: Any, output: torch.Tensor) -> Dict[str, torch.Tensor]:
+        batch_size, _, h, w = self._feature_shape
+        num_classes = output.shape[1]
+        predictions = output[:batch_size]
+        saliency_maps = output[batch_size:]
```

Comment:
It's a little confusing for me. What format (shape) do `output`, `predictions` and `saliency_maps` have?

Reply:
`output` is classifier head output so that it's B x C in general.
BTW, as we manipulated the feature map, B = B_org + B_org * H_feat * W_feat.
We split them to the actual prediction (B_org x C) and the saliency output (B_org * H_feat * W_feat x C), which will be transposed and reshaped to B_org x C x H_feat x W_feat.

Reply:
Thank you for the clarification! Can you leave a small comment regarding output structure. something like "Output contains combined resized predictions (bxc) and saliency maps (bxhxwc)"

Reply:
Coming in the next PR :)
![image](https://github.com/user-attachments/assets/cc0ed98f-3061-4f6e-ad2f-f46ca9b91304)